src/ln/channelmonitor.rs

```diff
@@ -170,34 +230,60 @@ impl<Key : Send + cmp::Eq + hash::Hash + 'static> SimpleManyChannelMonitor<Key>
 		let mut monitors = self.monitors.lock().unwrap();
 		match monitors.get_mut(&key) {
 			Some(orig_monitor) => {
-				log_trace!(self, "Updating Channel Monitor for channel {}", log_funding_option!(monitor.funding_txo));
+				log_trace!(self, "Updating Channel Monitor for channel {}", log_funding_option!(monitor.key_storage));
 				return orig_monitor.insert_combine(monitor);
 			},
 			None => {}
 		};
-		match &monitor.funding_txo {
-			&None => {
-				log_trace!(self, "Got new Channel Monitor for no-funding-set channel (monitoring all txn!)");
-				self.chain_monitor.watch_all_txn()
-			},
-			&Some((ref outpoint, ref script)) => {
-				log_trace!(self, "Got new Channel Monitor for channel {}", log_bytes!(outpoint.to_channel_id()[..]));
-				self.chain_monitor.install_watch_tx(&outpoint.txid, script);
-				self.chain_monitor.install_watch_outpoint((outpoint.txid, outpoint.index as u32), script);
+		match monitor.key_storage {
+			Storage::Local { ref funding_info, .. } => {
+				match funding_info {
+					&None => {
+						log_trace!(self, "Got new Channel Monitor for no-funding-set channel (monitoring all txn!)");
+						self.chain_monitor.watch_all_txn()
+					},
+					&Some((ref outpoint, ref script)) => {
+						log_trace!(self, "Got new Channel Monitor for channel {}", log_bytes!(outpoint.to_channel_id()[..]));
+						self.chain_monitor.install_watch_tx(&outpoint.txid, script);
+						self.chain_monitor.install_watch_outpoint((outpoint.txid, outpoint.index as u32), script);
+					},
+				}
 			},
+			Storage::Watchtower { .. } => {
+				unimplemented!();
```

Comment:
Why is this the unimplemented branch? Shouldn't the Local.funding_txo == None branch be the ignored one (since its nonsense to be adding a local monitor without a funding txo). Though of course it shouldnt panic, it should return an Err in that case.

Reply:
Add Err for Local.funding_txo == None, watch_all_txn for Watchtower